### PR TITLE
Contex.headers and Header type

### LINKED ISSUE #372: `Contex.headers` and `Header` type
- `Context.headers` method in `scontext` module;
- `SType::SHeader` type;
- `sheader` module with Header properties (see `scontext` as example);

### PR DESCRIPTION
**Pull request by SabaunT**
Close #372 and rework for https://github.com/ergoplatform/sigma-rust/pull/406

### CODE REVIEW

File: ergotree-ir/src/types/sheader.rs [hunk 1]
```diff
@@ -15,91 +17,165 @@ lazy_static! {
...
+            &MINER_PK_PROPERTY_METHOD_DESC,
+            &POW_ONETIME_PK_PROPERTY_METHOD_DESC,
+            &POW_NONCE_PROPERTY_METHOD_DESC,
+            &POW_DISTANCE_PROPERTY_METHOD_DESC,
+            &VOTES_PROPERTY_METHOD_DESC,
         ]
     ;
 }
 
+// todo all to one lazy_static
```
Comment (SabaunT) [inline ergotree-ir/src/types/sheader.rs:39]: @greenhat what do you think?

File: ergo-lib/src/chain/block_header.rs [hunk 2]
```diff
@@ -1,102 +1,136 @@
...
 #[derive(PartialEq, Eq, Debug, Clone)]
-pub struct Votes(pub [u8; 3]);
-
-/// Votes errors
-#[derive(Error, Debug)]
-pub enum VotesError {
-    /// Invalid byte array size
-    #[error("Votes: Invalid byte array size ({0})")]
-    InvalidSize(#[from] std::array::TryFromSliceError),
+pub(in crate::chain) struct BlockHeaderRef {
```
Comment (greenhat) [inline ergo-lib/src/chain/block_header.rs:29]: To keep naming consistent `HeaderSerde` should be a better fit. We already have a few "remote" types for serde with `Serde` suffix.

File: ergotree-ir/src/mir/value.rs [hunk 3]
```diff
@@ -158,6 +159,8 @@ pub enum Value {
     Tup(TupleItems<Value>),
     /// Transaction(and blockchain) context info
     Context,
+    /// Block header
+    Header(Header),
```
Comment (greenhat) [inline ergotree-ir/src/mir/value.rs:163]: Due to the Rust memory allocation for enum (always allocate by the biggest enum variant size) it needs to be boxed.

File: ergotree-ir/src/mir/value.rs [hunk 4]
```diff
@@ -376,6 +392,36 @@ impl<T: TryExtractFrom<Value> + StoreWrapped> TryExtractFrom<Value> for Vec<T> {
     }
 }
 
+impl<T: TryExtractFrom<Value> + StoreWrapped, const N: usize> TryExtractFrom<Value> for [T; N] {
```
Comment (greenhat) [inline ergotree-ir/src/mir/value.rs:395]: Nice!

File: ergotree-ir/src/types/sheader.rs (hunk 1 above)
Comment (greenhat) [reply ergotree-ir/src/types/sheader.rs:39]: I agree. Good call!

**Review by greenhat — APPROVED** [review]
Good job! Congratulations on your first contribution. The only thing that is missing is the evaluation for `Header` and `PreHeader` properties. But that should be taken care of in a separate PR.

I left a few notes, but I'm going to merge it now as-is and take care of them in my PR https://github.com/ergoplatform/sigma-rust/pull/397 which also moves a lot of types into `ergotree-ir`.

**Pull request merged**